packages/api/src/command/medical/patient/patient-import/patient-done.ts

```
@@ -0,0 +1,147 @@
+ *
+ * @param cxId - The customer ID.
+ * @param jobId - The bulk import job ID.
+ * @param patientId - The patient ID.
+ * @param status - The status of the patient import.
+ * @throws BadRequestError if no facility ID is provided and there's more than one facility for the customer.
+ */
+export async function finishSinglePatientImport({
+  cxId,
+  jobId,
+  patientId,
+  status,
+  reasonForDev,
+}: {
+  cxId: string;
+  jobId: string;
+  patientId: string;
+  status: PatientImportEntryStatus;
+  reasonForDev?: string | undefined;
+}): Promise<void> {
+  const { log } = out(
+    `finishSinglePatientImport - cxId ${cxId}, jobId ${jobId}, patientId ${patientId}`
+  );
+  log(`Finishing import for patient, status ${status}`);
+
+  try {
+    const [patientMapping] = await Promise.all([
+      fetchPatientMappingOrFail({ cxId, jobId, patientId }),
+      // validate access to the job
+      getPatientImportJobOrFail({ cxId, id: jobId }),
+    ]);
+    const rowNumber = patientMapping.rowNumber;
+
+    const patientRecord = await fetchPatientRecordOrFail({
+      cxId,
+      jobId,
+      rowNumber,
+    });
+
+    const updateTotalsOnDb = async (): Promise<FinishPatientImportParams | undefined> => {
+      const [[updatedRows]] = await PatientImportModel.increment(
+        [
+          ...(status === "successful" ? ["successful" as const] : []),
+          ...(status === "failed" ? ["failed" as const] : []),
+        ],
+        {
+          where: {
+            cxId,
+            id: jobId,
+          },
+          // Sequelize types are a mismatch, had to force this
+        } as IncrementDecrementOptionsWithBy<PatientImportModel>
+      );
+      // Using any because Sequelize doesn't map the columns to the model, even using mapToModel/model
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const updatedRaw = (updatedRows as unknown as any[] | undefined)?.[0];
+      if (!updatedRaw) return undefined;
```

maybe hardfail here? returning undefined to then have some weird "We shouldn't get here code" below seems unnecessary

---

see comment above: maybe move existence to getting-the-model-or-fail in parallel with get-patient-record-or-fail

---

The idea was to avoid error the whole job and instead treat errors while processing individual patients as something recoverable. But I agree, we can simplify and throw on missing response.

---

done